File: internal/validate/rules.go
package validate

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"

	"github.com/ekucher/bravo-bis-configurator/internal/schema"
)

// typeCoercionError checks that value can be interpreted as t, returning
// (message, true) if not. TypeString/TypePath/TypeEnum accept any string —
// enum membership is enforced separately by a RuleEnum, not by the type
// itself, since not every enum-typed field necessarily has a rule
// attached in every schema.
func typeCoercionError(t schema.FieldType, value string) (string, bool) {
	switch t {
	case schema.TypeInt:
		if _, err := strconv.ParseInt(strings.TrimSpace(value), 10, 64); err != nil {
			return fmt.Sprintf("value %q is not a valid integer", value), true
		}
	case schema.TypeFloat:
		if _, err := strconv.ParseFloat(strings.TrimSpace(value), 64); err != nil {
			return fmt.Sprintf("value %q is not a valid number", value), true
		}
	case schema.TypeBool:
		if _, err := parseBool(value); err != nil {
			return fmt.Sprintf("value %q is not a valid boolean (expected 0/1/true/false)", value), true
		}
	case schema.TypeString, schema.TypePath, schema.TypeEnum:
		// Any string is structurally valid; semantic checks are rule-driven.
	}
	return "", false
}

// parseBool accepts the value conventions actually observed in the real
// bravo.ini/bis.ini samples (0/1) plus the common true/false spelling, all
// case-insensitively.
func parseBool(value string) (bool, error) {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "0", "false":
		return false, nil
	case "1", "true":
		return true, nil
	default:
		return false, fmt.Errorf("not a boolean: %q", value)
	}
}

// checkRule runs r against value, returning (message, true) if it fails.
func checkRule(r *schema.ValidationRule, value string) (string, bool) {
	switch r.Kind {
	case schema.RulePathExists:
		return checkPathExists(r, value)
	case schema.RuleRegex:
		return checkRegex(r, value)
	case schema.RuleEnum:
		return checkEnum(r, value)
	case schema.RuleRange:
		return checkRange(r, value)
	default:
		return fmt.Sprintf("internal error: unknown validation rule kind %q", r.Kind), true
	}
}

func checkPathExists(r *schema.ValidationRule, value string) (string, bool) {
	if strings.TrimSpace(value) == "" {
		return "path is empty", true
	}
	info, err := os.Stat(value)
	if err != nil {
		return fmt.Sprintf("path does not exist: %s", value), true
	}
	switch r.PathMode {
	case schema.PathModeFile:
		if info.IsDir() {
			return fmt.Sprintf("expected a file but found a directory: %s", value), true
		}
	case schema.PathModeDir:
		if !info.IsDir() {
			return fmt.Sprintf("expected a directory but found a file: %s", value), true
		}
	case schema.PathModeEither, "":
		// no further check
	}
	return "", false
}

func checkRegex(r *schema.ValidationRule, value string) (string, bool) {
	re, err := regexp.Compile(r.Pattern)
	if err != nil {
		return fmt.Sprintf("internal error: invalid regex pattern %q: %v", r.Pattern, err), true
	}
	if !re.MatchString(value) {
		return fmt.Sprintf("value %q does not match required pattern %q", value, r.Pattern), true
	}
	return "", false
}

func checkEnum(r *schema.ValidationRule, value string) (string, bool) {
	for _, v := range r.Values {
		if v == value {
			return "", false
		}
	}
	return fmt.Sprintf("value %q is not one of the allowed values %v", value, r.Values), true
}

func checkRange(r *schema.ValidationRule, value string) (string, bool) {
	n, err := strconv.ParseFloat(strings.TrimSpace(value), 64)
	if err != nil {
		return fmt.Sprintf("value %q is not numeric, cannot check range", value), true
	}
	if r.Min != nil && n < *r.Min {
		return fmt.Sprintf("value %v is below the minimum %v", n, *r.Min), true
	}
	if r.Max != nil && n > *r.Max {
		return fmt.Sprintf("value %v is above the maximum %v", n, *r.Max), true
	}
	return "", false
}

File: src/BravoBisConfigurator.Core/Validate/Rules.cs
using System.Text.RegularExpressions;
using BravoBisConfigurator.Core.Schema;

namespace BravoBisConfigurator.Core.Validate;

/// <summary>
///  Type-coercion and ValidationRule checks. Ported 1:1 from
///  internal/validate/rules.go.
/// </summary>
internal static class Rules
{
    /// <summary>
    ///  Checks that value can be interpreted as t, returning (message, true)
    ///  if not. String/Path/Enum accept any string — enum membership is
    ///  enforced separately by a RuleEnum, not by the type itself, since not
    ///  every enum-typed field necessarily has a rule attached in every
    ///  schema.
    /// </summary>
    public static (string message, bool failed) TypeCoercionError(FieldType t, string value)
    {
        switch (t)
        {
            case FieldType.Int:
                if (!long.TryParse(value.Trim(), out _))
                {
                    return ($"value \"{value}\" is not a valid integer", true);
                }
                break;
            case FieldType.Float:
                if (!double.TryParse(value.Trim(), System.Globalization.NumberStyles.Float, System.Globalization.CultureInfo.InvariantCulture, out _))
                {
                    return ($"value \"{value}\" is not a valid number", true);
                }
                break;
            case FieldType.Bool:
                if (!TryParseBool(value, out _))
                {
                    return ($"value \"{value}\" is not a valid boolean (expected 0/1/true/false)", true);
                }
                break;
            case FieldType.String:
            case FieldType.Path:
            case FieldType.Enum:
                // Any string is structurally valid; semantic checks are rule-driven.
                break;
        }
        return ("", false);
    }

    /// <summary>
    ///  Accepts the value conventions actually observed in the real
    ///  bravo.ini/bis.ini samples (0/1) plus the common true/false spelling,
    ///  all case-insensitively.
    /// </summary>
    private static bool TryParseBool(string value, out bool result)
    {
        switch (value.Trim().ToLowerInvariant())
        {
            case "0":
            case "false":
                result = false;
                return true;
            case "1":
            case "true":
                result = true;
                return true;
            default:
                result = false;
                return false;
        }
    }

    /// <summary>Runs r against value, returning (message, true) if it fails.</summary>
    public static (string message, bool failed) CheckRule(ValidationRule r, string value) => r.Kind switch
    {
        RuleKind.PathExists => CheckPathExists(r, value),
        RuleKind.Regex => CheckRegex(r, value),
        RuleKind.Enum => CheckEnum(r, value),
        RuleKind.Range => CheckRange(r, value),
        _ => ($"internal error: unknown validation rule kind \"{r.Kind}\"", true),
    };

    private static (string, bool) CheckPathExists(ValidationRule r, string value)
    {
        if (value.Trim() == "")
        {
            return ("path is empty", true);
        }
        var isFile = File.Exists(value);
        var isDir = Directory.Exists(value);
        if (!isFile && !isDir)
        {
            return ($"path does not exist: {value}", true);
        }
        switch (r.PathMode)
        {
            case PathMode.File:
                if (isDir)
                {
                    return ($"expected a file but found a directory: {value}", true);
                }
                break;
            case PathMode.Dir:
                if (!isDir)
                {
                    return ($"expected a directory but found a file: {value}", true);
                }
                break;
        }
        return ("", false);
    }

    private static (string, bool) CheckRegex(ValidationRule r, string value)
    {
        Regex re;
        try
        {
            re = new Regex(r.Pattern ?? "");
        }
        catch (ArgumentException ex)
        {
            return ($"internal error: invalid regex pattern \"{r.Pattern}\": {ex.Message}", true);
        }
        if (!re.IsMatch(value))
        {
            return ($"value \"{value}\" does not match required pattern \"{r.Pattern}\"", true);
        }
        return ("", false);
    }

    private static (string, bool) CheckEnum(ValidationRule r, string value)
    {
        if (r.Values is not null && r.Values.Contains(value))
        {
            return ("", false);
        }
        var list = "[" + string.Join(" ", r.Values ?? new List<string>()) + "]"; // matches Go's %v on []string
        return ($"value \"{value}\" is not one of the allowed values {list}", true);
    }

    private static (string, bool) CheckRange(ValidationRule r, string value)
    {
        if (!double.TryParse(value.Trim(), System.Globalization.NumberStyles.Float, System.Globalization.CultureInfo.InvariantCulture, out var n))
        {
            return ($"value \"{value}\" is not numeric, cannot check range", true);
        }
        if (r.Min is { } min && n < min)
        {
            return ($"value {n} is below the minimum {min}", true);
        }
        if (r.Max is { } max && n > max)
        {
            return ($"value {n} is above the maximum {max}", true);
        }
        return ("", false);
    }
}
